cpp/src/arrow/type.h

```diff
@@ -451,40 +461,48 @@ class ARROW_EXPORT HalfFloatType
                                uint16_t> {
  public:
   Precision precision() const override;
-  std::string name() const override { return "halffloat"; }
+  static constexpr const char* type_name() { return "halffloat"; }
 };
 
 /// Concrete type class for 32-bit floating-point data (C "float")
 class ARROW_EXPORT FloatType
     : public detail::CTypeImpl<FloatType, FloatingPointType, Type::FLOAT, float> {
  public:
   Precision precision() const override;
-  std::string name() const override { return "float"; }
+  static constexpr const char* type_name() { return "float"; }
```

Comment:
Could this limit people that want to extend the code base in the future?

Reply:
I don't know, why would it?

Reply:
if name() actually had to contain non-trivial logic?  I wonder why it had this signature to begin with.

Reply:
I don't know, it seems the non-trivial logic ends up in ToString() by design.
As for the rationale behind `name()`, perhaps @wesm or @xhochy know about the history?

Reply:
I don't think the `name()` function is important. It's been around since the beginning of the project but I don't see it used much